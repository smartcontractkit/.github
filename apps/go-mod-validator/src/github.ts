import * as github from "@actions/github";
import * as core from "@actions/core";
import {
  BaseGoModule,
  GoModule,
  GoModuleWithCommitSha,
  GoModuleWithTag,
} from "./deps";

export type Octokit = ReturnType<typeof github.getOctokit>;

async function isCommitInDefaultBranch(
  gh: Octokit,
  defaultBranch: string,
  { repo, owner, commitSha }: GoModuleWithCommitSha,
): Promise<boolean> {
  const {
    data: { status },
  } = await gh.rest.repos.compareCommits({
    repo,
    owner,
    base: defaultBranch,
    head: commitSha,
  });

  return status === "identical" || status === "behind";
}

async function isTagInDefaultBranch(
  gh: Octokit,
  defaultBranch: string,
  mod: GoModuleWithTag,
): Promise<boolean> {
  const { data: repoTags } = await gh.rest.repos.listTags({
    owner: mod.owner,
    repo: mod.repo,
  });

  for (const repoTag of repoTags) {
    if (repoTag.name != mod.tag) {
      continue;
    }

    return isCommitInDefaultBranch(gh, defaultBranch, {
      ...mod,
      commitSha: repoTag.commit.sha,
    });
  }

  return false;
}

/**
 * Checks if a given go module in its respective GitHub repository's default branch.
 *
 * The version can be a tag or a commit.
 *
 * @param path - The path of the GitHub repository in the format "github.com/owner/repo".
 * @param version - The version to validate.
 * @param gh - The Octokit client used to make API requests to GitHub.
 *
 * @returns A boolean indicating whether the version exists in the repository.
 */
export async function isGoModReferencingDefaultBranch(
  mod: GoModule,
  defaultBranch: string,
  gh: Octokit,
) {
  core.debug(
    `Processing module: ${mod.name} ${mod.path} ${mod.version} ${mod.goModFilePath}`,
  );

  if ("commitSha" in mod) {
    return isCommitInDefaultBranch(gh, defaultBranch, mod);
  } else if ("tag" in mod) {
    return isTagInDefaultBranch(gh, defaultBranch, mod);
  } else {
    core.warning(`Unable to parse commit sha nor tag for module ${mod.name}`);
  }

  return false;
}

/**
 * Memoized function to get the default branch of a repository.
 */
export function defaultBranchGetter(gh: Octokit) {
  return async function getRepoDefaultBranch({ owner, repo }: BaseGoModule) {
    const resp = await gh.rest.repos.get({ owner, repo });

    return resp.data.default_branch;
  };
}
