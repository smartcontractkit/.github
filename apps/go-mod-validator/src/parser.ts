const smartContractKitPrefix = "github.com/smartcontractkit/";

function parseGoModLine(line: string) {
  // Trim any extra whitespace
  line = line.trim();

  // Go modules have psuedo versions and real versions
  // 1. psuedo version: v1.1.1-0.20190123174540-a2c9a5303de7
  // 2. real versions: v1.0.2
  // also, there can be pre-release versions
  // 1. v1.1.1-0.20190123174540-a2c9a5303de7+incompatible
  // 2. v1.0.2-beta
  // more details on - https://go.dev/doc/modules/version-numbers
  const moduleRegex = new RegExp(
    /^([\w.-]+(?:\/[\w.-]+)*) (v\d+\.\d+\.\d+(?:-[\w\d.-]+)?(?:\+[\w\d.-]+)?)$/,
  );
  const replaceRegex = new RegExp(
    moduleRegex.source.slice(0, moduleRegex.source.length - 1) +
      " => " +
      moduleRegex.source.slice(1),
  );
  const moduleMatch = line.match(moduleRegex);
  const replaceMatch = line.match(replaceRegex);

  if (moduleMatch) {
    const [, module, version] = moduleMatch;
    return {
      module,
      version,
    };
  }

  if (replaceMatch) {
    const [, , , module, version] = replaceMatch;
    return {
      module,
      version,
    };
  }

  return null;
}

// extract all final dependencies
function parseGoModLines(lines: string[]) {
  const dependencies: { module: string; version: string }[] = [];

  // exclude the first line which is module information
  lines = lines.slice(1);

  // parse each line to find out the final dependency
  lines.forEach((line) => {
    if (line.length == 0) {
      return;
    }

    const result = parseGoModLine(line);
    if (result) {
      console.debug(
        `Parsed Module: ${result.module}, Version: ${result.version}`,
      );
      dependencies.push(result);
    } else {
      console.error(`Could not parse line: ${line}`);
      throw new Error(`Could not parse line: ${line}`);
    }
  });

  return dependencies;
}

// extract only smartcontractkit dependencies
export function getSmartcontractkitDependencies(goListOutputContent: string) {
  // Split the content into lines
  const lines = goListOutputContent.split("\n");

  const dependencies: { module: string; version: string }[] = [];
  parseGoModLines(lines).forEach((dependency) => {
    if (dependency.module.startsWith(smartContractKitPrefix)) {
      dependencies.push(dependency);
    }
  });

  return dependencies;
}
