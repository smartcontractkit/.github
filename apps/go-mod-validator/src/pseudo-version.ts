// This file is a relatively direct translation of https://cs.opensource.google/go/x/mod/+/refs/tags/v0.20.0:module/pseudo.go
// Copyright (c) 2009 The Go Authors. All rights reserved.

// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are
// met:

//    * Redistributions of source code must retain the above copyright
// notice, this list of conditions and the following disclaimer.
//    * Redistributions in binary form must reproduce the above
// copyright notice, this list of conditions and the following disclaimer
// in the documentation and/or other materials provided with the
// distribution.
//    * Neither the name of Google Inc. nor the names of its
// contributors may be used to endorse or promote products derived from
// this software without specific prior written permission.

// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

import { valid } from "@snyk/go-semver";
import * as semver from "@snyk/go-semver/dist/go/semver";
const pseudoVersionRegex =
  /^v[0-9]+\.(0\.0-|\d+\.\d+-([^+]*\.)?0\.)\d{14}-[A-Za-z0-9]+(\+[0-9A-Za-z-]+(\.[0-9A-Za-z-]+)*)?$/;

export function isPseudoVersion(v: string): boolean {
  const hyphens = v.split("-").length;
  const validSemver = valid(v);
  const passesRegex = pseudoVersionRegex.test(v);

  return !!(hyphens >= 3 && validSemver && passesRegex);
}

export function pseudoVersionRev(v: string) {
  const { rev } = parsePseudoVersion(v);
  return rev;
}

function parsePseudoVersion(v: string): {
  base: string;
  timestamp: string;
  rev: string;
  build: string;
  err: Error | null;
} {
  if (!isPseudoVersion(v)) {
    throw new Error("pseudo syntax error");
  }

  const build = semver.build(v);
  v = v.replace(build, "");
  const j = v.lastIndexOf("-");
  const rev = v.slice(j + 1);
  v = v.slice(0, j);
  const i = v.lastIndexOf("-");

  if (v.lastIndexOf(".") > i) {
    const j = v.lastIndexOf(".");
    return {
      base: v.slice(0, j),
      timestamp: v.slice(j + 1),
      rev: rev,
      build: build,
      err: null,
    };
  } else {
    return {
      base: v.slice(0, i),
      timestamp: v.slice(i + 1),
      rev: rev,
      build: build,
      err: null,
    };
  }
}
