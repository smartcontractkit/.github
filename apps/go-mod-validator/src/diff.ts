import * as core from "@actions/core";
import { GetResponseTypeFromEndpointMethod } from "@octokit/types";
import { Octokit } from "./github";
type CompareResponse = GetResponseTypeFromEndpointMethod<
  Octokit["rest"]["repos"]["compareCommitsWithBasehead"]
>;
export type GithubFiles = CompareResponse["data"]["files"];

export async function getChangedGoModFiles(
  gh: Octokit,
  base: string,
  head: string,
  owner: string,
  repo: string,
  depPrefix: string,
): Promise<ParsedFile[]> {
  console.log({
    base,
    head,
    owner,
    repo
  })

  const files = await getComparison(gh, owner, repo, base, head);
  const relevantFiles = filterForRelevantChanges(files, );
  return parseAllAdditions(relevantFiles, depPrefix);
}

 function filterForRelevantChanges(files: GithubFiles): GithubFiles {
  return files?.filter(({ filename }) => filename.endsWith("go.mod"));
}


 async function getComparison(
  octokit: Octokit,
  owner: string,
  repo: string,
  base: string,
  head: string,
): Promise<GithubFiles> {
  core.debug(`Comparing ${owner}/${repo} commits ${base}...${head}`);

  const diff = await octokit.rest.repos.compareCommitsWithBasehead({
    owner,
    repo,
    basehead: `${base}...${head}`,
    // <before>...<after> or <earlier>...<later>
  });

  return diff.data.files;
}

// Taken from https://github.com/smartcontractkit/.github/blob/dc8b1a0b478151119d86ac1bf121ea7eb3c1c88c/actions/gha-workflow-validator/src/utils.ts#L59C1-L97C2
export interface ParsedFile {
  filename: string;
  sha: string;
  addedLines: FileAddition[];
}

export interface FileAddition {
  lineNumber: number;
  content: string;
}

 function parseAllAdditions(files: GithubFiles, depPrefix: string): ParsedFile[] {
  if (!files) return [];

  return files?.map((entry) => {
    const { filename, sha, patch } = entry;
    const addedLines = patch ? parsePatchAdditions(patch, depPrefix) : [];
    return { filename, sha, addedLines };
  });
}

 function parsePatchAdditions(patch: string, depPrefix: string): FileAddition[] {
  const lineChanges = patch?.split("\n") || [];

  const additions: FileAddition[] = [];

  let currentLineInFile = 0;
  for (const line of lineChanges) {
    if (line.startsWith("@@")) {
      // @@ denotes a git hunk header
      // https://mattscodecave.com/posts/howto-reading-git-diffs-and-staging-hunks.html
      // example line: @@ -16,10 +16,10 @@ jobs:
      //   - "-16,10": "-" denotes source file, "16,10" means the hunk starts at line 16 and output contains 10 lines
      //   - "+16,10": "+" denotes destination file, "16,10" means the same as above
      const [, , destination] = line.split(" ");

      if (!destination.startsWith("+")) {
        throw new Error("Invalid git hunk format");
      }

      const [destinationLine] = destination.substring(1).split(",");
      currentLineInFile = parseInt(destinationLine, 10);
      continue;
    } else if (line.startsWith("+")) {
      const currentLine = line.substring(1);
      if (isGoModule(currentLine, depPrefix)) {
        additions.push({
          content: currentLine,
          lineNumber: currentLineInFile,
        });
      }
    } else if (line.startsWith("-")) {
      // ignore deletions
      continue;
    }
    currentLineInFile++;
  }

  return additions;
}

function isGoModule(line: string, depPrefix: string): boolean {
  // Quick and dirty, we just search for github.com and include it in the results
  return line.includes(depPrefix);
}
