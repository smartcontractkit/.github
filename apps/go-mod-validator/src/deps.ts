import { execSync } from "child_process";
import { dirname, sep } from "path";
import { readFileSync } from "fs";
import * as core from "@actions/core";
import * as glob from "@actions/glob";
import { isPseudoVersion, pseudoVersionRev } from "./pseudo-version";

/**
 * Taken from `go help list` for the command `go list -json -m all`
 *  @link https://pkg.go.dev/cmd/go#hdr-List_packages_or_modules
 *  type Module struct {
 *        Path       string        // module path
 *        Query      string        // version query corresponding to this version
 *        Version    string        // module version
 *        Versions   []string      // available module versions
 *        Replace    *Module       // replaced by this module
 *        Time       *time.Time    // time version was created
 *        Update     *Module       // available update (with -u)
 *        Main       bool          // is this the main module?
 *        Indirect   bool          // module is only indirectly needed by main module
 *        Dir        string        // directory holding local copy of files, if any
 *        GoMod      string        // path to go.mod file describing module, if any
 *        GoVersion  string        // go version used in module
 *        Retracted  []string      // retraction information, if any (with -retracted or -u)
 *        Deprecated string        // deprecation message, if any (with -u)
 *        Error      *ModuleError  // error loading module
 *        Origin     any           // provenance of module
 *        Reuse      bool          // reuse of old module info is safe
 *    }
 *
 *    type ModuleError struct {
 *        Err string // the error itself
 *    }
 *
 */
interface GoMod {
  Path: string;
  Query: string;
  Version: string;
  Versions: string[];
  Replace?: GoMod;
  Time: Date | null;
  Update?: GoMod;
  Main: boolean;
  Indirect: boolean;
  Dir: string;
  GoMod: string;
  GoVersion: string;
  Retracted: string[];
  Deprecated: string;
  Error: ModuleError | null;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  Origin: any;
  Reuse: boolean;
}

interface ModuleError {
  Err: string;
}

export interface BaseGoModule {
  /**
   * The name of the module is the path + version of the module,
   * along with a postfix of "// indirect" indicating if the module is indirect.
   */
  name: string;
  /**
   * Module path
   */
  path: string;
  /**
   * Module version
   */
  version: string;

  /**
   * The repo owner of the dependency.
   */
  owner: string;

  /**
   * The repo name of the dependency.
   */
  repo: string;

  /**
   * The path to the go.mod file describing the module.
   *
   * Not to be confused with "path", which is the module path, like github.com/smartcontractkit/go-proxy.
   * @example /path/to/go.mod
   */
  goModFilePath: string;
}

export interface GoModuleWithCommitSha extends BaseGoModule {
  commitSha: string;
}

export interface GoModuleWithTag extends BaseGoModule {
  tag: string;
}

export type GoModule = GoModuleWithCommitSha | GoModuleWithTag | BaseGoModule;

function parseGoModListOutput(jsonStr: string): GoMod[] {
  /*
   * {
   *   "Path": "github.com/smartcontractkit/libocr",
   *   "Version": "v0.0.0-20240419185742-fd3cab206b2c"
   * }
   * {
   *   "Path": "github.com/smartcontractkit/go-proxy",
   *   "Version": "v0.1.0"
   * }
   */

  // store parsed objects
  const goModules: GoMod[] = [];

  // keep concatenating lines until the object is complete
  let objStr = "";

  // keep track of the depth of the object
  let objDepth = 0;

  // Process
  for (const l of jsonStr.split("\n")) {
    if (l.includes("{")) {
      objDepth++;
    }
    if (l.includes("}")) {
      objDepth--;
    }

    objStr += l;
    if (objDepth === 0 && objStr) {
      goModules.push(JSON.parse(objStr));
      objStr = "";
    }
  }

  return goModules;
}

/**
 * Recursively gets the paths of all go.mod files in the specified directory.
 *
 * @param goModDir The directory to search for go.mod files.
 *
 * @throws If no go.mod files are found or an error occurs.
 */
export async function getAllGoModsWithin(goModDir: string): Promise<string[]> {
  let files: string[] = [];
  const pattern = `${goModDir}/**/go.mod`;
  try {
    const globber = await glob.create(pattern);
    files = await globber.glob();
    files = files.map((f) => f.replace(`${process.cwd()}${sep}`, `.${sep}`));
  } catch (error) {
    throw new Error(`failed to get go.mod files: ${error}`);
  }

  if (files.length == 0) {
    throw new Error(
      `no go.mod files found with pattern ${pattern} and cwd ${process.cwd()}`,
    );
  }
  return files;
}

/**
 * Gets all the dependencies of the Go modules in the specified directory.
 *
 * We recursively collect all go.mod files within a directory, then return the flattened list of all dependencies.
 * There may be duplicates in the list, but we don't care about that.
 *
 * Dependencies are filtered by depPrefix.
 *
 * @param rootDir The directory containing the Go modules.
 *
 */
export async function getDeps(
  rootDir: string,
  depPrefix: string,
): Promise<GoModule[]> {
  const modFilePaths = await getAllGoModsWithin(rootDir);

  const deps = modFilePaths.flatMap((goModFilePath) => {
    core.info(`Finding dependencies in ${goModFilePath}`);
    try {
      const dir = dirname(goModFilePath);
      const output = execSync("go list -json -m all", {
        encoding: "utf-8",
        cwd: dir,
      });
      core.debug(`Raw output: ${output}`);

      const parsedDeps = parseGoModListOutput(output);
      core.debug(`Parsed dependencies: ${JSON.stringify(parsedDeps)}`);

      const mappedDeps = goModsToGoModules(
        goModFilePath,
        parsedDeps,
        depPrefix,
      );
      core.debug(`Mapped dependencies: ${JSON.stringify(mappedDeps)}`);

      return mappedDeps;
    } catch (error) {
      throw Error(
        `failed to get go.mod dependencies from file: ${goModFilePath}, err: ${error}`,
      );
    }
  });

  return deps;
}

/**
 * Creates a function that finds the line number of a dependency path in a go.mod file.
 */
export function lineForDependencyPathFinder() {
  const cache: {
    // goModPath -> [go mod file lines]
    [goModPath: string]: string[];
  } = {};

  /**
   * Given a go.mod file path and a dependency path, returns the line number of the dependency path in the go.mod file.
   *
   * We assume that you cannot have two modules of the same name in the go.mod file.
   * Technically, you can, but you need to make use of the replace directive along with two copies of the same module locally in order to do so.
   * This seems to be a rare case, so we are not handling it.
   *
   * @param goModPath The path to the go.mod file. E.g. /path/to/go.mod
   * @param depPath The dependency path to search for in the go.mod file. E.g. github.com/smartcontractkit/go-proxy
   *
   * @throws If the dependency path is found more than once in the go.mod file.
   */
  return function getDepPath({
    goModFilePath,
    path,
    name,
  }: BaseGoModule): number {
    core.debug(
      `Finding line number for ${path} in ${goModFilePath}. (${name})`,
    );
    if (!cache[goModFilePath]) {
      cache[goModFilePath] = readFileSync(goModFilePath, "utf-8")
        .split("\n")
        .map((l) => l.trim());
    }

    let line = -1;
    for (let i = 0; i < cache[goModFilePath].length; i++) {
      // HACK: We add a space after the depPath to avoid matching substrings.
      if (cache[goModFilePath][i].includes(path + " ")) {
        if (line !== -1) {
          core.warning(
            `Duplicate dependency path found: ${path} in ${goModFilePath} (line ${i + 1}). Previously found on line ${line}. Annotations may be faulty. Skipping.`,
          );
          continue;
        }
        line = i + 1;
      }
    }

    if (line === -1) {
      throw new Error(`dependency path not found: ${path}`);
    }

    return line;
  };
}

/**
 * Takes the parsed output of `go list -json -m all` and converts it to a list of GoModule objects.
 *
 * Performs the following transformations:
 * - Filters out the main module and non-org dependencies.
 * - Maps the GoMod object to a GoModule object.
 * @param goMods
 */
export function goModsToGoModules(
  goModFilePath: string,
  goMods: GoMod[],
  depPrefix: string,
): GoModule[] {
  const goModules = goMods
    // Replace the module with the Replace field if it exists
    .map((d) => d.Replace || d)
    // `go list -m -json all` also lists the main package, avoid parsing it.
    // and only validate dependencies belonging to our org
    .filter((d) => !d.Main && d.Path.startsWith(depPrefix))
    .map((d: GoMod): GoModule => {
      // repo format github.com/smartcontractkit/chainlink
      const [, owner, repo] = d.Path.split("/");
      const baseModule: BaseGoModule = {
        owner,
        repo,
        goModFilePath,
        name: `${d.Path}@${d.Version}${d.Indirect ? " // indirect" : ""}`,
        path: d.Path,
        version: d.Version,
      };

      const versionType = getVersionType(d.Version);

      if (!versionType.tag && !versionType.commitSha) {
        core.warning(
          `invalid version string: ${d.Version} for module: ${d.Path}`,
        );
        return baseModule;
      } else if (versionType.commitSha) {
        return {
          ...baseModule,
          commitSha: versionType.commitSha,
        };
      } else {
        return {
          ...baseModule,
          tag: versionType.tag,
        };
      }
    });

  return goModules;
}

/**
 * Extracts the type of a given version string.
 *
 * @param versionString - The version string to extract the type from.
 *
 * A pseudo version is a version string that is generated by Go when a module is built from a commit that is not tagged. Thus, they are generated from commit data.
 * See: https://go.dev/ref/mod#pseudo-versions for more details.
 *
 * A regular version is a version string that is a tag.
 *
 */
export function getVersionType(versionString: string) {
  // matches real versions like v0.1.0
  const versionRegex = /^(v\d+\.\d+\.\d+)$/;

  if (isPseudoVersion(versionString)) {
    return {
      commitSha: pseudoVersionRev(versionString),
      tag: undefined,
    };
  }
  const versionMatch = versionString.match(versionRegex);
  if (versionMatch) {
    return {
      commitSha: undefined,
      tag: versionMatch?.[1],
    };
  }

  return {
    commitSha: undefined,
    tag: undefined,
  };
}
