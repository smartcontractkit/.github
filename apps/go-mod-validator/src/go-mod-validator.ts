import { isGoModReferencingDefaultBranch } from "./github";
import { getDeps } from "./deps";
import { FIXING_ERRORS } from "./strings";
import * as github from "@actions/github";
import * as core from "@actions/core";
import { throttling } from "@octokit/plugin-throttling";

function getContext() {
  const goModDir = core.getInput("go-mod-dir", { required: true });
  const githubToken = core.getInput("github-token", { required: true });
  const depPrefix = core.getInput("dep-prefix", { required: true });
  const gh = github.getOctokit(
    githubToken,
    {
      throttle: {
        onRateLimit: (retryAfter, options, octokit, retryCount) => {
          octokit.log.warn(
            `Request quota exhausted for request ${options.method} ${options.url}`,
          );

          if (retryCount < 1) {
            // only retries once
            octokit.log.info(`Retrying after ${retryAfter} seconds!`);
            return true;
          }
        },
        onSecondaryRateLimit: (retryAfter, options, octokit) => {
          // does not retry, only logs a warning
          octokit.log.warn(
            `SecondaryRateLimit detected for request ${options.method} ${options.url}`,
          );
        },
      },
    },
    throttling,
  );

  return { goModDir, gh, depPrefix };
}

export async function run(): Promise<string> {
  const { goModDir, gh, depPrefix } = getContext();

  const depsToValidate = await getDeps(goModDir, depPrefix);

  // <dependency-name, error-string>
  const errs: Map<string, string> = new Map();

  const validating = depsToValidate.map(async (d) => {
    const isValid = await isGoModReferencingDefaultBranch(d, gh);

    if (!isValid) {
      errs.set(d.name, "dependency not on default branch");
    }
  });

  await Promise.all(validating);
  if (errs.size > 0) {
    errs.forEach((depName, errMsg) =>
      core.error(`validation failed for: ${errMsg}, err: ${depName}`),
    );

    core.summary.addRaw(FIXING_ERRORS, true);
    const summary = core.summary.stringify();
    await core.summary.write();

    core.setFailed(`validation failed for ${errs.size} dependencies`);
    return summary;
  } else {
    const msg = "validation successful for all go.mod dependencies";
    core.info(msg);
    return msg;
  }
}
