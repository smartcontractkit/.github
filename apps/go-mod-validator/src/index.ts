import { validateDependency } from "./github";
import { getDependenciesMap } from "./deps";

const smartContractKitPrefix = "github.com/smartcontractkit";

async function run() {
  // Read required env vars
  const githubToken = process.env.GITHUB_TOKEN || "";
  if (githubToken == "") {
    console.error("no GITHUB_TOKEN env variable found");
    process.exit(1);
  }

  // get dependencies from go.mod file
  const dependenciesMap = getDependenciesMap();

  // Verify each of the dependencies
  let validationErr = null;
  for (let [file, dependencies] of dependenciesMap.entries()) {
    console.info(`\nvalidating dependencies for ${file}`);

    for (let dependency of dependencies) {
      // handle replace redirectives
      if (dependency.Replace != undefined) {
        dependency = dependency.Replace;
      }

      // `go list -m -json all` also lists the main pacakge, avoid parsing it.
      // and only validate dependencies belonging to our org
      if (
        dependency.Version == undefined ||
        !dependency.Path.startsWith(smartContractKitPrefix)
      ) {
        continue;
      }

      try {
        if (
          await validateDependency(
            dependency.Path,
            dependency.Version,
            githubToken,
          )
        ) {
          console.info(
            `${dependency.Path}@${dependency.Version} is found in the default branch`,
          );
        } else {
          console.error(
            `${dependency.Path}@${dependency.Version} not found in the default branch`,
          );
          validationErr = new Error(
            `${dependency.Path}@${dependency.Version} not found in the default branch`,
          );
        }
      } catch (err) {
        console.error(
          `failed to verify dependency: ${dependency.Path}@${dependency.Version}, with err: ${err}`,
        );
        validationErr = err;
      }
    }
  }

  if (validationErr != null) {
    process.exit(1);
  }
}

run();
