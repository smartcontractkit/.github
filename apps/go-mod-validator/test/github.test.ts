import { isGoModReferencingDefaultBranch } from "../src/github";
import { describe, expect, it, vi } from "vitest";
import listTagsFixture from "./data/list_tags.json";
import { getVersionType, GoModule } from "../src/deps";

const owner = "smartcontractkit";
const repo = "go-plugin";
const goModPath = `github.com/${owner}/${repo}`;
const mockCompareCommits = vi.fn();
const mockListTags = vi.fn();
const mockOctokit: any = {
  rest: {
    repos: {
      compareCommits: mockCompareCommits,
      listTags: mockListTags,
    },
  },
};

// Mock data for tests
// https://cs.opensource.google/go/x/mod/+/refs/tags/v0.20.0:module/pseudo_test.go
// https://cs.opensource.google/go/x/mod/+/refs/tags/v0.20.0:module/pseudo.go;l=164
const pseudoTests = [
  { major: "", older: "", version: "v0.0.0-20060102150405-hash" },
  { major: "v0", older: "", version: "v0.0.0-20060102150405-hash" },
  { major: "v1", older: "", version: "v1.0.0-20060102150405-hash" },
  { major: "v2", older: "", version: "v2.0.0-20060102150405-hash" },
  { major: "unused", older: "v0.0.0", version: "v0.0.1-0.20060102150405-hash" },
  { major: "unused", older: "v1.2.3", version: "v1.2.4-0.20060102150405-hash" },
  {
    major: "unused",
    older: "v1.2.99999999999999999",
    version: "v1.2.100000000000000000-0.20060102150405-hash",
  },
  {
    major: "unused",
    older: "v1.2.3-pre",
    version: "v1.2.3-pre.0.20060102150405-hash",
  },
  {
    major: "unused",
    older: "v1.3.0-pre",
    version: "v1.3.0-pre.0.20060102150405-hash",
  },
  {
    major: "unused",
    older: "v0.0.0--",
    version: "v0.0.0--.0.20060102150405-hash",
  },
  {
    major: "unused",
    older: "v1.0.0+metadata",
    version: "v1.0.1-0.20060102150405-hash+metadata",
  },
  {
    major: "unused",
    older: "v2.0.0+incompatible",
    version: "v2.0.1-0.20060102150405-hash+incompatible",
  },
  {
    major: "unused",
    older: "v2.3.0-pre+incompatible",
    version: "v2.3.0-pre.0.20060102150405-hash+incompatible",
  },
  {
    major: "unused",
    older: "unused",
    version: "v0.0.5-0.20220116011046-fa5810519dcb",
  },
];

describe.only("getVersionType", () => {
  for (const { version } of pseudoTests) {
    it("should return the correct version type", () => {
      const verType = getVersionType(version);
      expect(verType).toMatchSnapshot();
    });
  }

  it("should parse out the git sha correctly", () => {
    let version = "v0.2.2-0.20240808143317-6b16fc28887d";
    let verType = getVersionType(version);
    expect(verType).toEqual({ commitSha: "6b16fc28887d", tag: undefined });

    version = "v0.0.1-beta-test.0.20240709043547-03612098f799";
    verType = getVersionType(version);
    expect(verType).toEqual({ commitSha: "03612098f799", tag: undefined });
  });
});

describe("isGoModReferencingDefaultBranch", () => {
  const fullRepo = {
    owner,
    repo,
  };

  describe("pseudo-versions", () => {
    const commitSha = "b3b91517de16";
    const version = `v0.0.0-20240208201424-${commitSha}`;
    const goMod: GoModule = {
      ...fullRepo,
      path: goModPath,
      version,
      commitSha,
      name: `${goModPath}@${version}`,
      goModFilePath: ".",
    };

    it("should check if the commit is present on the default branch", async () => {
      mockCompareCommits.mockResolvedValueOnce({
        data: {
          status: "behind",
        },
      });

      const referencesDefault = await isGoModReferencingDefaultBranch(
        goMod,
        "main",
        mockOctokit,
      );

      expect(mockListTags).to.not.toHaveBeenCalled();
      expect(mockCompareCommits).toHaveBeenCalledWith({
        owner,
        repo,
        base: "main",
        head: commitSha,
      });
      expect(referencesDefault).toBeTruthy();
    });

    it("should throw an error if the compare commits request fails", async () => {
      const errMessage = "not found";
      mockCompareCommits.mockRejectedValue(errMessage);

      const isReferencingDefaultBranch = () =>
        isGoModReferencingDefaultBranch(goMod, "main", mockOctokit);
      expect(mockListTags).to.not.toHaveBeenCalled();
      await expect(isReferencingDefaultBranch).rejects.toThrowError(errMessage);
    });
  });

  describe("regular versions", () => {
    const version = "v0.1.0";
    const goMod: GoModule = {
      ...fullRepo,
      path: goModPath,
      version,
      tag: version,
      name: `${goModPath}@${version}`,
      goModFilePath: ".",
    };

    it("should check if the commit is present on the default branch", async () => {
      mockCompareCommits.mockResolvedValueOnce({
        data: {
          status: "behind",
        },
      });
      mockListTags.mockResolvedValueOnce({
        data: listTagsFixture,
      });

      const expectedCommitSha = listTagsFixture.find(
        (tag) => tag.name === version,
      )?.commit.sha;
      expect(expectedCommitSha).toBeDefined();

      const isValid = await isGoModReferencingDefaultBranch(
        goMod,
        "main",
        mockOctokit,
      );
      expect(mockListTags).toHaveBeenCalledOnce();
      expect(mockCompareCommits).toHaveBeenCalledWith({
        owner,
        repo,
        base: "main",
        head: expectedCommitSha,
      });
      expect(isValid).toEqual(true);
    });

    it("should throw an error if listing tags fails", async () => {
      const errMessage = "not found";
      mockListTags.mockRejectedValue(errMessage);

      const result = isGoModReferencingDefaultBranch(goMod, "", mockOctokit);
      await expect(result).rejects.toThrow(errMessage);
    });
  });
});
