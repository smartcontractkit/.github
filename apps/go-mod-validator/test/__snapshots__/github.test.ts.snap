// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getVersionType > should return the correct version type 1`] = `
{
  "commitSha": undefined,
  "tag": undefined,
}
`;

exports[`getVersionType > should return the correct version type 2`] = `
{
  "commitSha": undefined,
  "tag": undefined,
}
`;

exports[`getVersionType > should return the correct version type 3`] = `
{
  "commitSha": undefined,
  "tag": undefined,
}
`;

exports[`getVersionType > should return the correct version type 4`] = `
{
  "commitSha": undefined,
  "tag": undefined,
}
`;

exports[`getVersionType > should return the correct version type 5`] = `
{
  "commitSha": undefined,
  "tag": undefined,
}
`;

exports[`getVersionType > should return the correct version type 6`] = `
{
  "commitSha": undefined,
  "tag": undefined,
}
`;

exports[`getVersionType > should return the correct version type 7`] = `
{
  "commitSha": undefined,
  "tag": undefined,
}
`;

exports[`getVersionType > should return the correct version type 8`] = `
{
  "commitSha": undefined,
  "tag": undefined,
}
`;

exports[`getVersionType > should return the correct version type 9`] = `
{
  "commitSha": undefined,
  "tag": undefined,
}
`;

exports[`getVersionType > should return the correct version type 10`] = `
{
  "commitSha": undefined,
  "tag": undefined,
}
`;

exports[`getVersionType > should return the correct version type 11`] = `
{
  "commitSha": undefined,
  "tag": undefined,
}
`;

exports[`getVersionType > should return the correct version type 12`] = `
{
  "commitSha": undefined,
  "tag": undefined,
}
`;

exports[`getVersionType > should return the correct version type 13`] = `
{
  "commitSha": undefined,
  "tag": undefined,
}
`;

exports[`getVersionType > should return the correct version type 14`] = `
{
  "commitSha": "fa5810519dcb",
  "tag": undefined,
}
`;
